keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,505 @@
+      "hugepages-1Gi": "0",
+      "hugepages-2Mi": "0",
+      "memory": "196146004Ki",
+      "pods": "110"
+    },
+    "allocatable": {
+      "cpu": "58",
+      "exclusive-cpu": "54",
+      "ephemeral-storage": "767528359485",
+      "hugepages-1Gi": "0",
+      "hugepages-2Mi": "0",
+      "memory": "186067796Ki",
+      "pods": "110"
+    },
+  ...
+```
+
+In kube-scheduler, `ExlusiveMilliCPU` is added in scheduler's `Resource` structure and `NodeResourcesFit` plugin is extended to filter out nodes that can not meet pod's exclusive CPU request.
+
+A new item `ExclusiveMilliCPU` is added in the scheduler `Resource` structure:
+
+```
+// Resource is a collection of compute resource.
+type Resource struct {
+        MilliCPU          int64
+        ExclusiveMilliCPU int64    // added
+        Memory            int64
+        EphemeralStorage  int64
+        // We store allowedPodNumber (which is Node.Status.Allocatable.Pods().Value())
+        // explicitly as int, to avoid conversions and improve performance.
+        AllowedPodNumber int
+        // ScalarResources
+        ScalarResources map[v1.ResourceName]int64
+}
+```
+
+A new node fitting failure 'Insufficient exclusive cpu' is added in the `NodeResourcesFit` plugin:
+
+```
+        if podRequest.MilliCPU > 0 && podRequest.MilliCPU > (nodeInfo.Allocatable.MilliCPU-nodeInfo.Requested.MilliCPU) {
+                insufficientResources = append(insufficientResources, InsufficientResource{
+                        ResourceName: v1.ResourceCPU,
+                        Reason:       "Insufficient cpu",
+                        Requested:    podRequest.MilliCPU,
+                        Used:         nodeInfo.Requested.MilliCPU,
+                        Capacity:     nodeInfo.Allocatable.MilliCPU,
+                })
+        }
+        if nodeInfo.Allocatable.ExclusiveMilliCPU > 0 {    // added
+                if podRequest.ExclusiveMilliCPU > 0 && podRequest.ExclusiveMilliCPU > (nodeInfo.Allocatable.ExclusiveMilliCPU-nodeInfo.Requested.ExclusiveMilliCPU) {
+                        insufficientResources = append(insufficientResources, InsufficientResource{
+                                ResourceName: v1.ResourceExclusiveCPU,
+                                Reason:       "Insufficient exclusive cpu",
+                                Requested:    podRequest.ExclusiveMilliCPU,
+                                Used:         nodeInfo.Requested.ExclusiveMilliCPU,
+                                Capacity:     nodeInfo.Allocatable.ExclusiveMilliCPU,
+                        })
+                }
+        }
+```
+
+#### Archived Risk Mitigation (Option 2)
+
+Another proposal brought up is to force the cpu requests for best effort pods to 1 MilliCPU in kubelet. This option is meant to be simpler than option-1, but the different resource accounting in kubelet and kube-scheduler can create runaway pods similar to that in https://github.com/kubernetes/kubernetes/issues/84869.
```

Comment:
let's just mention for historical purpose that we agreed to start with the metrics to assess the actual impact in real deployment of the concern before to revisit if we need risk mitigation.